### ISSUE TITLE
Differing component implementation logic across recipes

### ISSUE THREAD

**Issue by EugenHotaj** [pytorch/torchtune#2307]
I've noticed that torchtune has different implementations for some components across recipes. A couple of examples:

Gradient accumulation:
* `full_distributed_finetune.py` correctly computes the loss across ranks / gas ([here](https://github.com/pytorch/torchtune/blob/main/recipes/full_finetune_distributed.py#L777-L792)).
* `lora_dpo_distributed.py` (and #2275) use the "old" but incorrect implementation ([here](https://github.com/pytorch/torchtune/blob/main/recipes/lora_dpo_distributed.py#L713)).

Tokens per second:
* `full_distributed_finetune.py` computes `tps` by considering the unmasked tokens only.
*  `lora_dpo_distributed.py` computes `tps` by considering all tokens.

There could be more but these are the main ones that jumped out. 

It would be awesome for torchtune to have more uniformity across the recipes! At least some of the key ones that are highly used (sft, dpo, maybe ppo). Right now it can get a bit confusing when switching recipes and seeing metrics you don't expect.

**Comment by felipemello1** [comment]
hey @EugenHotaj , 100% agreed. As we added new functionalities, it became harder to keep all recipes 100% aligned. 

We have as top priority for this year to reduce complexity in recipes and make them easier to maintain. One way to help with that, for example, is to moduralize some functionalities and make them utilities. Another way is to increase testing coverage so the burden of testing is reduced (e.g. combinations of quantization, lora, distributed, compile, performance flags, dataset packed, etc). We also plan on investing more on RL this year.

For now, please continue raising tickets / submitting PRs as you find issues like that. Thanks for all of your contributions!

**Comment by EugenHotaj** [comment]
@SalmanMohammadi no worries, it's not urgent, just wanted to raise awareness of the issue / hope it can get prioritized in the future 🙂

**Comment by Ankur-singh** [comment]
@SalmanMohammadi I have been going through the recipes and had a similar experience. Can we create a new issue (and add "community help wanted" tag) to track all such differences in one place? Something similar to #1810.

Since this is a high priority this year, it makes sense to track and tackle everything in one place. Plus, I can take a crack at a few myself!
